### DISCUSSION TITLE
How to use alteration on a finish workflow instance?

### DISCUSSION THREAD

**Discussion by quytrieu37** [elsa-workflows/elsa-core#5747] [Q&A]
Hi everyone,

I'm using alteration to schedule an activity on my long-running workflow with the following code:
```
var alterationPlan = new ScheduleActivity { ActivityId = activityId };

var results = await _alterationRunner.RunAsync(instanceId, new List<IAlteration>() { alterationPlan }, CancellationToken.None);

await _workflowDispatcher.DispatchAsync(results, CancellationToken.None);
```

The goal is to continue a workflow if an exception occurs. This works well with a suspended workflow, where the workflow continues at the provided `activityId`.

However, when I try this on a finished workflow, the `results` object returns with the `IsSuccess` flag set to true, but nothing happens.

I have a few questions related to this:

1. How can I apply an alteration to a finished workflow instance to schedule an activity and continue the process? Do I need to change the workflow status or something similar?
2. Is there a way to start a workflow, one that has not been run before, from any activity, not just from the starting point?

Any help will be appreciated! Thank you so much!

**Comment by glime-ai** [comment]
To address your questions:

1. Alterations are typically used to modify the state of running or suspended workflow instances. Once a workflow instance is finished, its execution has concluded, and it cannot be altered to resume execution. If you need to perform additional work after a workflow has finished, you might consider starting a new workflow instance or using a different mechanism to perform the required tasks.

2. Starting a workflow from any activity other than the starting point is not a typical feature of workflow engines, including Elsa. Workflows are generally designed to start from a defined initial state and then proceed through their activities based on the workflow definition. If you need to start execution from a specific point, you might need to design your workflow to allow for this by using conditional logic or by defining multiple workflows that represent different starting points.

If you need to retry specific activities within a workflow due to exceptions, you should handle this within the workflow's execution logic, using try-catch blocks or compensation activities to manage exceptions and perform retries as necessary.

For more advanced scenarios, you might want to look into customizing the workflow execution behavior by implementing custom activities or handlers.

Please refer to the Elsa documentation and resources for more information on handling workflow execution and alterations:

- [Introduction to Alterations](https://v3.elsaworkflows.io/docs/alterations/introduction)
- [Running Alterations](https://v3.elsaworkflows.io/docs/alterations/run-alterations)
- [Alteration Plans API](https://v3.elsaworkflows.io/docs/alterations/alteration-plans-api)
- [Issue on GitHub discussing retry with `ActivityInstanceId`](https://github.com/elsa-workflows/elsa-core/issues/4939)

If you encounter a specific issue or limitation with the Elsa workflow engine, consider reaching out to the community or opening an issue on the [Elsa GitHub repository](https://github.com/elsa-workflows/elsa-core).

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by quytrieu37** [comment]
For question 2, use `StartWorkflowRuntimeParams` with TriggerActivityId set to desire ActivityId to start workflow at specified activity:
```
var startParams = new StartWorkflowRuntimeParams
{
    TriggerActivityId = desireActivityId
};
await workflowRuntime.StartWorkflowAsync(request.DefinitionId, startParams);
```
